### DISCUSSION TITLE
work with labstack's echo web framework has problem

### DISCUSSION THREAD

**Discussion by wakaka6** [tmaxmax/go-sse#41] [Q&A] [answered]
Thank you for creating this amazing library!

```
package main

import (
	"errors"
	"log"
	"net/http"
	"time"

	"github.com/google/uuid"
	"github.com/labstack/echo/v4"
	"github.com/labstack/echo/v4/middleware"
	"github.com/tmaxmax/go-sse"
)

func main() {
	e := echo.New()
	s := &sse.Server{}

	go func() {

		for range time.Tick(time.Second) {
			// Update
			ev := &sse.Message{}
			time.Sleep(3 * time.Second)
			ev.AppendData("time: " + time.Now().Format(time.RFC3339Nano) + uuid.New().String())
			_ = s.Publish(ev)
		}
	}()

	e.Use(middleware.Logger())
	e.Use(middleware.Recover())

	e.GET("/sse", func(c echo.Context) error {
		log.Printf("SSE client connected, ip: %v", c.RealIP())

		ticker := time.NewTicker(1 * time.Second)
		defer ticker.Stop()

		go func() {
			<-c.Request().Context().Done() // Received Browser Disconnection
			log.Printf("The client is disconnected: %v\n", c.RealIP())
			return
		}()

		s.ServeHTTP(c.Response(), c.Request())
		return nil
	})

	if err := e.Start(":8080"); err != nil && !errors.Is(err, http.ErrServerClosed) {
		log.Fatal(err)
	}
}
```
I used this code although it works, echo gives me a warning, but when I use the r3labs/sse library in combination with echo I don't have this problem. So I can't determine if this is an echo problem or a go-sse library problem, can you help me locate it.
```
echo: http: superfluous response.WriteHeader call from github.com/labstack/echo/v4.(*Response).WriteHeader (response.go:65)
```

----
working with r3labs/sse code
https://echo.labstack.com/docs/cookbook/sse#using-3rd-party-library-r3labssse-to-broadcast-events

**Comment by tmaxmax** [comment, marked as answer]
Hi there, thank you for opening this discussion!

I've investigated this and found the cause: internally, `go-sse` flushes the request without explicitly writing the header, which triggers an implicit 200 OK write. This implicit write though doesn't toggle some internal Echo flags which then causes Echo to write the header again, hence the "superfluous" write. This is benign and you don't need to worry.

You shouldn't have see this warning every time, though – it is annoying and may distract attention from important warnings. Here's a way you can integrate `go-sse` with `echo` without such warnings (used your code as starting point):
```go
package main

import (
	"errors"
	"fmt"
	"log"
	"net/http"
	"time"

	"github.com/google/uuid"
	"github.com/labstack/echo/v4"
	"github.com/labstack/echo/v4/middleware"
	"github.com/tmaxmax/go-sse"
)

func main() {
	e := echo.New()
	s := &sse.Joe{}

	go func() {
		for range time.Tick(time.Second) {
			// Update
			ev := &sse.Message{}
			time.Sleep(3 * time.Second)
			ev.AppendData("time: " + time.Now().Format(time.RFC3339Nano) + uuid.New().String())
			_ = s.Publish(ev, []string{sse.DefaultTopic})
		}
	}()

	e.Use(middleware.Logger())
	e.Use(middleware.Recover())

	e.GET("/sse", func(c echo.Context) error {
		log.Printf("SSE client connected, ip: %v", c.RealIP())

		go func() {
			<-c.Request().Context().Done() // Received Browser Disconnection
			log.Printf("The client is disconnected: %v\n", c.RealIP())
		}()

		sess, err := sse.Upgrade(c.Response(), c.Request())
		if err != nil {
			c.Error(fmt.Errorf("initialize SSE connection: %w", err))
		}

		// Let Echo write the header – this is what removes the warning.
		// If you remove this line the warning will reappear.
		c.Response().WriteHeader(http.StatusOK)

		return s.Subscribe(c.Request().Context(), sse.Subscription{
			Client:      sess,
			LastEventID: sess.LastEventID,
			Topics:      []string{sse.DefaultTopic},
		})
	})

	if err := e.Start(":8080"); err != nil && !errors.Is(err, http.ErrServerClosed) {
		log.Fatal(err)
	}
}
```
Note the replacement of `sse.Server` with `sse.Joe` – that is directly the provider. Given that you use Echo as your HTTP handler you can skip the `sse.Server` completely, given that its sole role is to provide a `net/http.Handler`. You also don't need the `sse.Server.Logger` and `sse.Server.OnSubscribe` configuration properties, as you're writing the request handler now and you can add there any code you wish.

If unfamiliar with what a `go-sse` provider is or any other concept mentioned here, refer to the docs.

I'll most probably modify the library code in such a way that this warning won't occur regardless of usage, as for now I wouldn't want to document or create integrations with specific libraries, as it is very easy to do them. It might be valuable though to mention `sse.Upgrade` in the README.

Anyway, let me know if this helps and if I can assist you further with anything else! Apologies for the delayed response.
